{{cookiecutter.repostory_name}}/app/src/{{cookiecutter.django_project_name}}/{{cookiecutter.django_default_app_name}}/constance_prometheus.py

```diff
@@ -0,0 +1,124 @@
+"""
+Export django-constance variables as Prometheus metrics.
+
+By default this module exports all constance variables which are of the ``int``,
+``float``, ``bool`` or ``decimal.Decimal`` types and not included in the
+``CONSTANCE_PROMETHEUS_BLACKLIST`` sequence.
+
+The variables will appear named ``constance_config_VARIABLE_KEY`` and represented
+as Gauge metrics.
+
+The ``CONSTANCE_PROMETHEUS_WHITELIST`` sequence may specify additional variables
+to be exported. Their values will be cast into ``str`` and appended to the metric
+name (e.g. ``VAR="foo bar"`` will become ``constance_config_VAR_foo_bar``) and
+get 1.0 value. Once changed, the old metric becomes 0.0 and new one will appear,
+allowing to present different values on a graph.
+
+"""
+import os
+import re
+
+from constance import config, settings as constance_settings
+from constance.signals import config_updated
+from decimal import Decimal
+from django.conf import settings
+from django.dispatch import receiver
+from prometheus_client import REGISTRY, CollectorRegistry, Gauge
+
+from . import metrics
+
+__all__ = ("Metric", "on_constance_config_updated", "export_config")
+
+
+def is_blacklisted(key):
+    return key in getattr(settings, "CONSTANCE_PROMETHEUS_BLACKLIST", [])
+
+
+def is_whitelisted(key):
+    return key in getattr(settings, "CONSTANCE_PROMETHEUS_WHITELIST", [])
+
+
+def gets_monitored(key, value):
+    return not is_blacklisted(key) and (
+        isinstance(value, (int, float, bool, Decimal)) or is_whitelisted(key)
+    )
+
+
+METRIC_NAME_FILTER = re.compile(r"\W", re.ASCII)
+
+
+class Metric:
+    prefix = "constance_config"
+    config_key = None
+
+    def __init__(self, config_key=None):
+        self.config_key = config_key or self.config_key
+        self.name = self._get_name()
+        self.description = constance_settings.CONFIG[self.config_key][1]
+        value = getattr(config, self.config_key)
+        if isinstance(value, (bool, int, float, Decimal)):
+            self._metric = Gauge(self.name, self.description)
+            self.store = self._metric.set
+        else:
+            # a str or custom type; cast everything to str
+            value = str(value)
+            self.name = self._get_name(value)
+            self._metric = Gauge(self.name, self.description)
+            self.store = self._store_str
+
+    def _get_name(self, value=None):
+        n = f"{self.prefix:s}_{self.config_key:s}"
+        if value is not None:
+            v = METRIC_NAME_FILTER.sub('_', value)
+            n = f"{n}_{v}"
+        return n
+
+    def _unregister(self):
```

Comment:
@mpnowacki-reef I think you might know what to do here. I mean for me it's pretty clear that unregistering should be implemented in `metrics.py` and not in `constance_prometheus.py`, registry seems to be a singleton so ideally we'd see some CollectorRegistryProvider which would just return the exiting object, rather than the class magically returning the same object as created somewhere else or the class somehow working consistently across threads despite having multiple copies...

But other than that I don't know how to deal with the multiprocess unregistering.

Reply:
registry seems to not be a singleton, according to prometheus-client howto: https://github.com/prometheus/client_python#multiprocess-mode-eg-gunicorn


I'm a bit lost, it is my understanding that `registry.unregister` is for unregistering Collectors, not individual metrics, what ma I missing?

Reply:
My question is: does it actually work? When collecting metrics in multiprocessing mode, their values are read from .db files. I don't se how `unregister` would remove entries from those files but I can be wrong. 

Frankly speaking, from what I see, removing metrics does not really work when in multiprocessing mode. But again, I could be wrong. This whole thing is a little bit spaghetti in prometheus-client. Removing labels, however, should work, so a single metric with labels for constance param name and value could be a quasi-supported solution.

Reply:
I think it's kind of fixed right now. I say _kind_ because 15cd1fc is nothing but a hack. It seems to work in my local test project but will likely bite our collective ass in the future. What do you think?